<p>Most Android-powered devices have built-in sensors that measure motion, orientation, and various environmental conditions.
    In addition to these are the image sensor (a.k.a Camera) and the geopositioning sensor (a.k.a GPS).</p>
<br>The common point of all these sensors is that they are expensive while in use. Their common bug is to let the sensor
unnecessarily process data when the app enters an idle state, typically when paused or stopped.</br>
Consequently, calls must be carefully pairwised: <code>CLLocationManager.startUpdatingLocation() and CLLocationManager.stopUpdatingLocation()</code>.</br>
Failing to do so can drain the battery in just a few hours.</p>
<h2>Noncompliant Code Example</h2>
<pre>
import CoreLocation

class LocationTracker: NSObject, CLLocationManagerDelegate {
    var locationManager: CLLocationManager?

    override init() {
        super.init()
        locationManager = CLLocationManager()
        locationManager?.delegate = self
        locationManager?.requestAlwaysAuthorization() // Request appropriate authorization
        locationManager?.startUpdatingLocation() // Start location updates
    }

    // LocationManager Delegate Methods
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        // Process new locations
    }
}
</pre>
<br>
<h2>Compliant Code Example</h2>
<pre>
class LocationTracker: NSObject, CLLocationManagerDelegate {
    var locationManager: CLLocationManager?

    override init() {
        super.init()
        locationManager = CLLocationManager()
        locationManager?.delegate = self
        locationManager?.requestAlwaysAuthorization() // Request appropriate authorization
        locationManager?.startUpdatingLocation() // Start location updates only when needed
    }

    // LocationManager Delegate Methods
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        // Process new locations
        // Possibly stop updates if they are no longer needed
        locationManager?.stopUpdatingLocation()
    }
}
</pre>