<img src="http://www.neomades.com/extern/partage/ecoCode/2sur5_1x.png">
<p>Most iOS devices have built-in sensors that measure motion, orientation, and various environmental conditions.
    Additionally, they have image sensors (a.k.a. Camera) and geo-positioning sensors (a.k.a. GPS).

    The common point of all these sensors is that they consume significant power while in use. Their common issue is
    processing data unnecessarily when the app is in an idle state, typically when it enters the background or becomes
    inactive.

    <code>Consequently, calls to start and stop sensor updates must be carefully managed for motion sensor:
        CMMotionManager#startAccelerometerUpdates()/CMMotionManager#stopAccelerometerUpdates().
        Failing to do so can drain the battery quickly.</code></p>
<h2>Noncompliant Code Example</h2>
<pre>
import CoreMotion

let motionManager = CMMotionManager()

func startMotionUpdates() {
    if motionManager.isAccelerometerAvailable {
        motionManager.startAccelerometerUpdates(to: .main) { data, error in
            // Handle accelerometer updates
        }
    }
}
</pre>
<h2>Compliant Code Example</h2>
<pre>
import CoreMotion

let motionManager = CMMotionManager()

func startMotionUpdates() {
    if motionManager.isAccelerometerAvailable {
        motionManager.startAccelerometerUpdates(to: .main) { data, error in
            // Handle accelerometer updates
        }
    }
}

func stopMotionUpdates() {
    if motionManager.isAccelerometerActive {
        motionManager.stopAccelerometerUpdates()
    }
}
</pre>